### PR TITLE
Apply box adaptation when checking overrides

### PR DESCRIPTION
**Pull request by Linyxus**
In this PR, we allow box adaptation when checking the compatibility of overriding pairs in capture checking phase. For example, it allows the following code to compile:
```scala
class IO

abstract class A[X] {
  def foo(x: Unit): X
  def bar(op: X => Int): Int
}

class C

def test(io: {*} IO) = {
  class B extends A[{io} C] {  // X =:= {io} C
    def foo(x: Unit): {io} C = ???
    def bar(op: ({io} C) => Int): Int = 0
  }
}
```
The `foo` and `bar` in `B` are both valid overrides, since the types in `B` and `A[{io} C]` is compatible with box adaptation.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -290,8 +297,11 @@ object RefChecks {
    *  TODO check that classes are not overridden
    *  TODO This still needs to be cleaned up; the current version is a straight port of what was there
    *       before, but it looks too complicated and method bodies are far too large.
+   *
+   *   @param makeOverridePairsChecker  A function for creating a OverridePairsChecker instance
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:301]: I think we might be able to streamline this, maybe by moving checkOverrides into OverridingPairsChecker. But we can do that later after this PR is merged.

**Review by odersky — COMMENTED** [review]
Otherwise LGTM. Let's discuss the changes in baseType when I am back next week.

**Comment by odersky** [comment]
test performance please

**Comment by Linyxus** [comment]
From the test results it seems that this PR won't slow the compiler down. I just force pushed to fix one typo in the comments. Shall we get it merged? @odersky

**Comment by odersky** [comment]
Yes, ready to merge.

**Review by odersky — APPROVED** [review]
Looks good  now!

**Pull request merged**